### PR TITLE
match AST structures

### PR DESCRIPTION
**Pull request by ZacLN**
Just a heads up that this is my intention for future changes to CSTParser. Fields of `EXPR` are a superset of those of `Core.Expr` with `.head` and `.args` being directly equivalent between the two types. All other info stored optionally. 

Why? This allows (with a bit of delicate handling) for code that manipulation expressions to work across the two formats. 

Todo: 
- [x] getfield macro name handling with `@` at front.
- [x] full test coverage across all `head` fields.
- [x] document differences from AST.

EDIT:
@domluna, this will be of interest

### CODE REVIEW

**Comment by davidanthoff** [comment]
All post-Juliacon stuff, right?

**Comment by ZacLN** [comment]
Yes

**Comment by domluna** [comment]
@ZacLN could you expand a little more about what this does? Are there things that would be possible this that are not possible currently?

**Comment by ZacLN** [comment]
Hiya, this change produces `EXPR` that are identical to `Expr` that come from the flisp parser in terms of the `.head` and `.args` fields. All additional info (i.e. spans, other atomic expressions such as punctuation) are more clearly partitioned away. 

What are the implications?
* Removes the need to understand two different storage structures for expressions.
* There has been on and off talk of trying to integrate this parser into Base - this change would make that feasible.
* This is a necessary step for attempting to handle macros within the StaticLint/LanguageServer framework. (e.g. exploring an approach of interpreting macros)
* 2x speed up on parsing. (Not sure why, this wasn't an intended outcome).
* Iteration across expressions

**Comment by ZacLN** [comment]
@pfitzseb , this is ready for you to have a look. Let me know if you need a quick call to talk you through things.

**Comment by pfitzseb** [comment]
On it. Will probably not finish reviewing this before the holidays though.

**Comment by ZacLN** [comment]
That fits my timetable!

**Review by pfitzseb — COMMENTED** [review]
Ok, so I've tried to look through all the changes here (but couldn't be bothered to validate _all_ of the logic here, sorry :D). I have a few minor changes in https://github.com/julia-vscode/CSTParser.jl/pull/215.

On a high level I have two comments, mainly:
1. The `@closer`/`@nocloser` macros are confusing. Is there any reason they aren't implemented as closures? Would also be nice to have e.g. `closer_mod(close=(:comma, :inwhere), noclose=(:newline,), ps) do ... end` instead of `@nocloser(ps, :inwhere, @nocloser(ps, :newline, @closer(ps, :comma, ...)`, imho. This definitely shouldn't hold up this PR though, given that the macro design is a couple of years old at this point.
2. I don't really like the `iterate` implementation here. Do you think performance would suffer too much from storing `trivia` as basically a `index => trivia_at_that_index` dict or something like that? Feels like that'd simplify a bunch of code.

File: src/components/operators.jl
Comment (ZacLN) [reply src/components/operators.jl:51]: I'm not quite sure which ones you're querying?

Comment (pfitzseb) [reply src/components/operators.jl:51]: `<:`/`>:`.

**Comment by ZacLN** [comment]
Thanks for having a look. Agree on 1., but will leave that to a later date. 

On iteration - yes it's a bit grim and far too much code but I'm not sure there is a performant alternative. I'm not entirely clear what your suggestion is, could you elaborate a bit? I'm open to alternatives but I do think there's a need to store args and trivia in separate places which limits the possibilities somewhat.

**Comment by pfitzseb** [comment]
So currently we store `(1,2,3)` as something like
```
head: tuple
args: [1,2,3]
trivia: ["(", ",", ",", ")"]
```
and then iterate properly. I'd propose something like
```
head: tuple
args: [1,2,3]
trivia: Dict(0 => ["("], 1 => [","], 2 => [","], 3 => [")"])
```
That's easy to construct because the index is always just `length(args)` when constructing and makes iteration trivial because you can just iterate over `args` and then check if there's corresponding `trivia` (the inner vector might not even be necessary).

**Comment by ZacLN** [comment]
Ah, I see what you mean (though I think your example keys for the Dict above are wrong?). Instinctively I'm concerned about the overhead of replacing the vector with a Dict for each expression that has trivia - possibly quite costly. A quick and dirty test suggests c. +60% storage requirement (for the trivia) but I'll have a play.

**Comment by pfitzseb** [comment]
Do we just want to merge this as-is and potentially switch to a different implementation later on?

**Comment by davidanthoff** [comment]
I'm primarily in favor of anything that allows us to move quickly on to the binary serialization PR over at SymbolServer :)

**Comment by ZacLN** [comment]
Yes, I've not had a chance to give this any more time so lets move now on this and make any needed changes later.

**Review by pfitzseb — APPROVED** [review]

**Comment by pfitzseb** [comment]
Alright, I'll clean up https://github.com/julia-vscode/CSTParser.jl/pull/215 and reopen it against master tomorrow (it has at least one proper bug fix and imho the testing against Base is better there too).

**Comment by ZacLN** [comment]
Arg, accidentally pushed another small commit.

Anyway, https://github.com/julia-vscode/StaticLint.jl/pull/203, https://github.com/julia-vscode/LanguageServer.jl/pull/856, and https://github.com/julia-vscode/DocumentFormat.jl/pull/116 are the relevant partner PR's. I'll need to tidy them up before merging

**Review by pfitzseb — APPROVED** [review]

**Review by pfitzseb — APPROVED** [review]

**Comment by o314** [comment]
Very good work.

I have however two con :

- if you target a 100% compat with Expr, you will dispose of Julia made 10 (15?) years ago
- considering the breaking changes introduced by a not 100% compatible Expr2 in Base;
the cost of removing femtolisp, its runtime, its extension, doing macro expansion fully in julia; rebooting a clean and well ordered workflow upto eval will far exceed the cost of updating the Expr handling stuff

**Pull request merged**
